### PR TITLE
Added ModbusBridge

### PR DESCRIPTION
**Pull request by jeroenst**
## Description:
Implementation of ModbusBridge
https://github.com/arendst/Tasmota/discussions/15703

I coded this the best I can. Probably things are not as they are preferred. Please let me know how and what to change and I will change the code accordingly. 

When this PR is accepted I will start adding documentation.

## Checklist:
  - [x] The pull request is done against the latest development branch
  - [x] Only relevant files were touched
  - [x] Only one feature/fix was added per PR and the code change compiles without warnings
  - [x] The code change is tested and works with Tasmota core ESP8266 V.2.7.4.9
  - [ ] The code change is tested and works with Tasmota core ESP32 V.2.0.4 (I don't have this)
  - [x] I accept the [CLA](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md#contributor-license-agreement-cla).

_NOTE: The code change must pass CI tests. **Your PR cannot be merged unless tests pass**_

Tested and working ok:
```
09:42:26.311 CMD: modbussend {"deviceaddress": 1, "functioncode": 3, "startaddress": 1, "type":"raw", "count":4}
09:42:26.313 SRC: WebConsole from 172.30.30.100
09:42:26.315 CMD: Grp 0, Cmd 'MODBUSSEND', Idx 1, Len 83, Pld -99, Data '{"deviceaddress": 1, "functioncode": 3, "startaddress": 1, "type":"raw", "count":4}'
09:42:26.319 ORNO: ModbusBridge Sending: Addr:1 Func:3 Reg:1 Length:4 Type:8 Count:4
09:42:26.323 RSL: RESULT = {"ModbusSend":"Done"}
09:42:28.501 WIF: Checking connection...
09:42:38.511 DMP: 01 03 04 FF 00 FE 00 8A 47
09:42:38.512 ORNO: MODBUSBRIDGE received data. Length=9
09:42:38.516 RSL: RESULT = {"ModbusReceived":[1,3,4,255,0,254,0,138,71]}}
09:42:44.663 CMD: modbussend {"deviceaddress": 1, "functioncode": 3, "startaddress": 1, "type":"bit8", "count":4}
09:42:44.665 SRC: WebConsole from 172.30.30.100
09:42:44.668 CMD: Grp 0, Cmd 'MODBUSSEND', Idx 1, Len 84, Pld -99, Data '{"deviceaddress": 1, "functioncode": 3, "startaddress": 1, "type":"bit8", "count":4}'
09:42:44.672 ORNO: ModbusBridge Sending: Addr:1 Func:3 Reg:1 Length:4 Type:9 Count:4
09:42:44.676 RSL: RESULT = {"ModbusSend":"Done"}
09:42:45.761 DMP: 01 03 04 FF 00 FE 00 8A 47
09:42:45.763 ORNO: MODBUSBRIDGE received data. Length=9
09:42:45.767 RSL: RESULT = {"ModbusReceived":{"DeviceAddress":1,"FunctionCode":3,"StartAddress":1,"Length":9,"Count":4,"Values":[11111111,00000000,11111110,00000000]}}
09:42:48.505 WIF: Checking connection...
09:42:51.569 CMD: modbussend {"deviceaddress": 1, "functioncode": 3, "startaddress": 1, "type":"int8", "count":4}
09:42:51.571 SRC: WebConsole from 172.30.30.100
09:42:51.574 CMD: Grp 0, Cmd 'MODBUSSEND', Idx 1, Len 84, Pld -99, Data '{"deviceaddress": 1, "functioncode": 3, "startaddress": 1, "type":"int8", "count":4}'
09:42:51.580 ORNO: ModbusBridge Sending: Addr:1 Func:3 Reg:1 Length:4 Type:4 Count:4
09:42:51.584 RSL: RESULT = {"ModbusSend":"Done"}
09:42:53.012 DMP: 01 03 04 FF 00 FE 00 8A 47
09:42:53.014 ORNO: MODBUSBRIDGE received data. Length=9
09:42:53.018 RSL: RESULT = {"ModbusReceived":{"DeviceAddress":1,"FunctionCode":3,"StartAddress":1,"Length":9,"Count":4,"Values":[-1,0,-2,0]}}
09:42:57.460 CMD: modbussend {"deviceaddress": 1, "functioncode": 3, "startaddress": 1, "type":"int16", "count":2}
09:42:57.463 SRC: WebConsole from 172.30.30.100
09:42:57.465 CMD: Grp 0, Cmd 'MODBUSSEND', Idx 1, Len 85, Pld -99, Data '{"deviceaddress": 1, "functioncode": 3, "startaddress": 1, "type":"int16", "count":2}'
09:42:57.471 ORNO: ModbusBridge Sending: Addr:1 Func:3 Reg:1 Length:4 Type:5 Count:2
09:42:57.476 RSL: RESULT = {"ModbusSend":"Done"}
09:42:58.761 DMP: 01 03 04 FF 00 FE 00 8A 47
09:42:58.763 ORNO: MODBUSBRIDGE received data. Length=9
09:42:58.767 RSL: RESULT = {"ModbusReceived":{"DeviceAddress":1,"FunctionCode":3,"StartAddress":1,"Length":9,"Count":2,"Values":[65280,65024]}}
09:43:04.163 CMD: modbussend {"deviceaddress": 1, "functioncode": 3, "startaddress": 1, "type":"int32", "count":1}
09:43:04.166 SRC: WebConsole from 172.30.30.100
09:43:04.168 CMD: Grp 0, Cmd 'MODBUSSEND', Idx 1, Len 85, Pld -99, Data '{"deviceaddress": 1, "functioncode": 3, "startaddress": 1, "type":"int32", "count":1}'
09:43:04.173 ORNO: ModbusBridge Sending: Addr:1 Func:3 Reg:1 Length:4 Type:6 Count:1
09:43:04.177 RSL: RESULT = {"ModbusSend":"Done"}
09:43:05.266 DMP: 01 03 04 FF 00 FE 00 8A 47
09:43:05.268 ORNO: MODBUSBRIDGE received data. Length=9
09:43:05.271 RSL: RESULT = {"ModbusReceived":{"DeviceAddress":1,"FunctionCode":3,"StartAddress":1,"Length":9,"Count":1,"Values":[-16712192]}}
09:43:06.454 WIF: Sending Gratuitous ARP
09:43:08.513 WIF: Checking connection...
09:43:10.829 CMD: modbussend {"deviceaddress": 1, "functioncode": 3, "startaddress": 1, "type":"uint32", "count":1}
09:43:10.831 SRC: WebConsole from 172.30.30.100
09:43:10.834 CMD: Grp 0, Cmd 'MODBUSSEND', Idx 1, Len 86, Pld -99, Data '{"deviceaddress": 1, "functioncode": 3, "startaddress": 1, "type":"uint32", "count":1}'
09:43:10.839 ORNO: ModbusBridge Sending: Addr:1 Func:3 Reg:1 Length:4 Type:3 Count:1
09:43:10.843 RSL: RESULT = {"ModbusSend":"Done"}
09:43:11.762 DMP: 01 03 04 FF 00 FE 00 8A 47
09:43:11.764 ORNO: MODBUSBRIDGE received data. Length=9
09:43:11.768 RSL: RESULT = {"ModbusReceived":{"DeviceAddress":1,"FunctionCode":3,"StartAddress":1,"Length":9,"Count":1,"Values":[4278255104]}}
09:43:18.507 CMD: modbussend {"deviceaddress": 1, "functioncode": 3, "startaddress": 1, "type":"uint16", "count":2}
09:43:18.509 SRC: WebConsole from 172.30.30.100
09:43:18.512 CMD: Grp 0, Cmd 'MODBUSSEND', Idx 1, Len 86, Pld -99, Data '{"deviceaddress": 1, "functioncode": 3, "startaddress": 1, "type":"uint16", "count":2}'
09:43:18.518 ORNO: ModbusBridge Sending: Addr:1 Func:3 Reg:1 Length:4 Type:2 Count:2
09:43:18.522 RSL: RESULT = {"ModbusSend":"Done"}
09:43:20.512 DMP: 01 03 04 FF 00 FE 00 8A 47
09:43:20.514 ORNO: MODBUSBRIDGE received data. Length=9
09:43:20.518 RSL: RESULT = {"ModbusReceived":{"DeviceAddress":1,"FunctionCode":3,"StartAddress":1,"Length":9,"Count":2,"Values":[65280,65024]}}
09:43:28.525 WIF: Checking connection...
09:43:30.182 CMD: modbussend {"deviceaddress": 1, "functioncode": 3, "startaddress": 1, "type":"uint8", "count":4}
09:43:30.185 SRC: WebConsole from 172.30.30.100
09:43:30.188 CMD: Grp 0, Cmd 'MODBUSSEND', Idx 1, Len 85, Pld -99, Data '{"deviceaddress": 1, "functioncode": 3, "startaddress": 1, "type":"uint8", "count":4}'
09:43:30.193 ORNO: ModbusBridge Sending: Addr:1 Func:3 Reg:1 Length:4 Type:1 Count:4
09:43:30.198 RSL: RESULT = {"ModbusSend":"Done"}
09:43:31.266 DMP: 01 03 04 FF 00 FE 00 8A 47
09:43:31.268 ORNO: MODBUSBRIDGE received data. Length=9
09:43:31.272 RSL: RESULT = {"ModbusReceived":{"DeviceAddress":1,"FunctionCode":3,"StartAddress":1,"Length":9,"Count":4,"Values":[255,0,254,0]}}
```

### CODE REVIEW

**Comment by jeroenst** [comment]
Why are checks canceled and why is TasmotaModbus not found on ESP32 builds?

File: platformio.ini [hunk 1]
```diff
@@ -35,10 +35,10 @@ custom_unpack_dir           = unpacked_littlefs
 build_unflags               = ${core.build_unflags}
 build_flags                 = ${core.build_flags}
 monitor_speed               = 115200
-monitor_port                = COM5
+monitor_port                = /dev/ttyUSB0
```
Comment (arendst) [inline platformio.ini:38]: Revert these changes. No changes allowed in this file

File: tasmota/include/tasmota_template.h [hunk 2]
```diff
@@ -49,6 +49,7 @@ enum UserSelectablePins {
   GPIO_PMS5003_TX, GPIO_PMS5003_RX,    // Plantower PMS5003 Serial interface
   GPIO_SDS0X1_TX, GPIO_SDS0X1_RX,      // Nova Fitness SDS011 Serial interface
   GPIO_SBR_TX, GPIO_SBR_RX,            // Serial Bridge Serial interface
+  GPIO_MBR_TX, GPIO_MBR_RX,            // Modbus Bridge Serial interface
```
Comment (arendst) [inline tasmota/include/tasmota_template.h:52]: Move these to the end of the list as documented at the top of the list

File: tasmota/include/tasmota_template.h [hunk 3]
```diff
@@ -289,6 +290,7 @@ const char kSensorNames[] PROGMEM =
   D_SENSOR_PMS5003_TX "|" D_SENSOR_PMS5003_RX "|"
   D_SENSOR_SDS0X1_TX "|" D_SENSOR_SDS0X1_RX "|"
   D_SENSOR_SBR_TX "|" D_SENSOR_SBR_RX "|"
+  D_SENSOR_MBR_TX "|" D_SENSOR_MBR_RX "|"
```
Comment (arendst) [inline tasmota/include/tasmota_template.h:293]: Ditto

File: tasmota/include/tasmota_template_legacy.h [hunk 4]
```diff
@@ -471,7 +471,9 @@ const uint16_t kGpioConvert[] PROGMEM = {
   AGPIO(GPIO_IEM3000_TX),     // IEM3000 Serial interface
   AGPIO(GPIO_IEM3000_RX),     // IEM3000 Serial interface
   AGPIO(GPIO_ZIGBEE_RST),     // Zigbee reset
-  AGPIO(GPIO_DYP_RX)
+  AGPIO(GPIO_DYP_RX),
```
Comment (arendst) [inline tasmota/include/tasmota_template_legacy.h:474]: No changes allowed in this file. Pls remove.

File: tasmota/my_user_config.h [hunk 5]
```diff
@@ -782,6 +783,7 @@
 //#define USE_VINDRIKTNING                         // Add support for IKEA VINDRIKTNING particle concentration sensor (+0k6 code)
 //  #define VINDRIKTNING_SHOW_PM1                  // Display undocumented/supposed PM1.0 values
 //  #define VINDRIKTNING_SHOW_PM10                 // Display undocumented/supposed PM10 values
+#define USE_MODBUS_BRIDGE
```
Comment (arendst) [inline tasmota/my_user_config.h:786]: Remove this duplicate

File: tasmota/tasmota_xdrv_driver/xdrv_100_modbus_bridge.ino [hunk 6]
```diff
@@ -0,0 +1,455 @@
...
+#define D_JSON_MODBUS_FUNCTION_CODE "FunctionCode"
+#define D_JSON_MODBUS_START_ADDRESS "StartAddress"
+#define D_JSON_MODBUS_COUNT "Count"
+#define D_JSON_MODBUS_ENDIAN "Endian"
+#define D_JSON_MODBUS_TYPE "Type" // allready defined
+#define D_JSON_MODBUS_VALUES "Values"
+#define D_JSON_MODBUS_LENGTH "Length"
+
+const char kModbusBridgeCommands[] PROGMEM = "|" // No prefix
+    D_CMND_MODBUS_SEND "|" D_CMND_MODBUS_SETBAUDRATE "|" D_CMND_MODBUS_SETSERIALCONFIG;
```
Comment (arendst) [inline tasmota/tasmota_xdrv_driver/xdrv_100_modbus_bridge.ino:49]: Use prefix ``Modbus`` and remove from command names

File: tasmota/tasmota_xdrv_driver/xdrv_100_modbus_bridge.ino [hunk 7]
```diff
@@ -0,0 +1,455 @@
...
+    modbusBridge.deviceAddress = 0;
+    free(buffer);
+  }
+}
+
+/********************************************************************************************/
+
+void ModbusBridgeInit(void)
+{
+  tasmotaModbus = new TasmotaModbus(Pin(GPIO_MBR_RX), Pin(GPIO_MBR_TX));
```
Comment (arendst) [inline tasmota/tasmota_xdrv_driver/xdrv_100_modbus_bridge.ino:257]: Do not ussume users have configured these GPIOs. First check if they did and if not break out. See other drivers how to do this.

File: tasmota/tasmota_xdrv_driver/xdrv_100_modbus_bridge.ino [hunk 8]
```diff
@@ -0,0 +1,455 @@
...
+  {
+    if (2 == result)
+    {
+      AddLog(LOG_LEVEL_DEBUG, PSTR("ORNO: ModbusBridge HW serial init 8E1 at %d baud"), MBR_SPEED);
+      Serial.begin(MBR_SPEED, SERIAL_8E1);
+      ClaimSerial();
+    }
+    else
+    {
+      AddLog(LOG_LEVEL_DEBUG, PSTR("ORNO: ModbusBridge SW serial init 8E1 at %d baud"), MBR_SPEED);
```
Comment (arendst) [inline tasmota/tasmota_xdrv_driver/xdrv_100_modbus_bridge.ino:269]: Once you have all previous comments solved clean up duplicate messages. They take up a lot of code space especially on ESP8266

Make other messages a lot smaller and/or cryptic. You're not writing a book.

File: tasmota/tasmota_xdrv_driver/xdrv_100_modbus_bridge.ino [hunk 9]
```diff
@@ -0,0 +1,455 @@
...
+  }
+}
+
+/*********************************************************************************************\
+ * Commands
+\*********************************************************************************************/
+
+void CmndModbusBridgeSend(void)
+{
+  RemoveSpace(XdrvMailbox.data); // TODO is this really needed?
```
Comment (arendst) [inline tasmota/tasmota_xdrv_driver/xdrv_100_modbus_bridge.ino:280]: Nope

File: tasmota/tasmota_xdrv_driver/xdrv_100_modbus_bridge.ino (hunk 6 above)
Comment (jeroenst) [reply tasmota/tasmota_xdrv_driver/xdrv_100_modbus_bridge.ino:49]: I used https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_xdrv_driver/xdrv_08_serial_bridge.ino as an example, I don't understand what I have to do for this.

Comment (jeroenst) [reply tasmota/tasmota_xdrv_driver/xdrv_100_modbus_bridge.ino:49]: Do you want me to rename them to:

D_MODBUS_LENGTH
D_MODBUS_SEND
D_MODBUS_SETBAUDRATE

?

**Comment by jeroenst** [comment]
Sorry, something is going very wrong with git when trying to add the changes you proposed

File: tasmota/tasmota_xdrv_driver/xdrv_100_modbus_bridge.ino (hunk 6 above)
Comment (jeroenst) [reply tasmota/tasmota_xdrv_driver/xdrv_100_modbus_bridge.ino:49]: Ok thank you!

I will try to finish documentation this week.

File: platformio.ini (hunk 1 above)
Comment (arendst) [inline platformio.ini:38]: Do not change to /dev/ttyUSB0

File: platformio.ini [hunk 10]
```diff
@@ -35,10 +35,10 @@ custom_unpack_dir           = unpacked_littlefs
 build_unflags               = ${core.build_unflags}
 build_flags                 = ${core.build_flags}
 monitor_speed               = 115200
-monitor_port                = COM5
+monitor_port                = /dev/ttyUSB0
 ; *** Upload Serial reset method for Wemos and NodeMCU
 upload_resetmethod          = nodemcu
-upload_port                 = COM5
+upload_port                 = /dev/ttyUSB0
```
Comment (arendst) [inline platformio.ini:41]: Ditto

File: tasmota/include/tasmota_template.h (hunk 2 above)
Comment (arendst) [reply tasmota/include/tasmota_template.h:52]: Not resolved

File: tasmota/include/tasmota_template_legacy.h (hunk 4 above)
Comment (arendst) [reply tasmota/include/tasmota_template_legacy.h:474]: Still not resolved

Comment (jeroenst) [reply tasmota/include/tasmota_template_legacy.h:474]: See https://github.com/arendst/Tasmota/pull/16014

File: platformio.ini (hunk 1 above)
Comment (jeroenst) [reply platformio.ini:38]: See https://github.com/arendst/Tasmota/pull/16014

File: platformio.ini (hunk 10 above)
Comment (jeroenst) [reply platformio.ini:41]: See https://github.com/arendst/Tasmota/pull/16014
